datafusion/physical-plan/src/memory.rs

```diff
@@ -365,8 +365,165 @@ impl RecordBatchStream for MemoryStream {
     }
 }
 
+pub trait StreamingBatchGenerator: Send + Sync + fmt::Debug + fmt::Display {
+    /// Generate the next batch, return `None` when no more batches are available
+    fn generate_next_batch(&mut self) -> Result<Option<RecordBatch>>;
+
+    /// Creates a boxed clone of this generator.
+    ///
+    /// This method is required because `Clone` cannot be directly implemented for
+    /// trait objects. It provides a way to clone trait objects of
+    /// StreamingBatchGenerator while maintaining proper type erasure.
+    fn clone_box(&self) -> Box<dyn StreamingBatchGenerator>;
+}
+
+/// Execution plan for streaming in-memory batches of data
+///
+/// This plan generates output batches lazily, it doesn't have to buffer all batches
+/// in memory up front (compared to `MemoryExec`), thus consuming constant memory.
+pub struct StreamingMemoryExec {
+    /// Schema representing the data
+    schema: SchemaRef,
+    /// Functions to generate batches for each partition
+    batch_generators: Vec<Box<dyn StreamingBatchGenerator>>,
+    /// Total number of rows to generate for statistics
+    cache: PlanProperties,
+}
+
+impl StreamingMemoryExec {
+    /// Create a new streaming memory execution plan
+    pub fn try_new(
+        schema: SchemaRef,
+        generators: Vec<Box<dyn StreamingBatchGenerator>>,
+    ) -> Result<Self> {
+        let cache = PlanProperties::new(
+            EquivalenceProperties::new(Arc::clone(&schema)),
+            Partitioning::RoundRobinBatch(generators.len()),
```

Comment:
Is this field mean: Let's say a exec only have 1 generator, the optimizer will try to insert a `RepartitionExec` to the output of `StreamingMemoryExec`, and use round robin to repartition output batches to `target_partitions` number?
I found it works on some aggregate queries, but not for a sort query
```
> explain select * from generate_series(1, 10000) as t1(v1) order by v1;
+---------------+----------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                           |
+---------------+----------------------------------------------------------------------------------------------------------------+
| logical_plan  | Sort: t1.v1 ASC NULLS LAST                                                                                     |
|               |   SubqueryAlias: t1                                                                                            |
|               |     Projection: tmp_table.value AS v1                                                                          |
|               |       TableScan: tmp_table projection=[value]                                                                  |
| physical_plan | SortExec: expr=[v1@0 ASC NULLS LAST], preserve_partitioning=[false]                                            |
|               |   ProjectionExec: expr=[value@0 as v1]                                                                         |
|               |     StreamingMemoryExec: partitions=1, batch_generators=[generate_series: start=1, end=10000, batch_size=8192] |
|               |                                                                                                                |
+---------------+----------------------------------------------------------------------------------------------------------------+
```
I want to know is it correct here, and this sort query in theory can be made parallelized by changing somewhere in optimization phase? 🤔

Reply:
Maybe you can look at https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html#method.repartitioned or similar methods 🤔

Reply:
I'm not sure about this too